modules/ml/src/main/java/org/apache/ignite/ml/trainers/TrainerTransformers.java

```diff
@@ -0,0 +1,282 @@
+
+        for (int i = 0; i < ensembleSize; i++) {
+            tasks.add(trainingTaskGenerator.apply(bootstrappedBuilder, i, mappings != null ? extractors.get(i) : extractor));
+        }
+
+        List<ModelWithEndomorphism<Vector, Double, M>> models = environment.parallelismStrategy().submit(tasks)
+            .stream()
+            .map(Promise::unsafeGet)
+            .map(ModelWithEndomorphism<Vector, Double, M>::new)
+            .collect(Collectors.toList());
+
+        // If we need to do projection, do it.
+        if (mappings != null) {
+            for (int i = 0; i < models.size(); i++) {
+                models.get(i).setEndo(getProjector(mappings.get(i)));
+            }
+        }
+
+        double learningTime = (double) (System.currentTimeMillis() - startTs) / 1000.0;
+        log.log(MLLogger.VerboseLevel.LOW, "The training time was %.2fs.", learningTime);
+        log.log(MLLogger.VerboseLevel.LOW, "Learning finished.");
+
+        return new ModelsComposition(models, aggregator);
+    }
+
+    @NotNull
+    public static Map<Integer, Integer> getMapping(int featuresVectorSize, int maximumFeaturesCntPerMdl) {
+        int[] featureIdxs = Utils.selectKDistinct(featuresVectorSize, maximumFeaturesCntPerMdl, new Random());
+        Map<Integer, Integer> featureMapping = new HashMap<>();
+        IntStream.range(0, maximumFeaturesCntPerMdl)
+            .forEach(localId -> featureMapping.put(localId, featureIdxs[localId]));
+        return featureMapping;
+    }
+
+    /**
+     * Repeats each entry count of times distributed according given Poisson distribution.
+     *
+     * @param en                  Upstream entry.
+     * @param poissonDistribution Poisson distribution.
+     * @param <K>                 Type of keys of upstream data.
+     * @param <V>                 Type of values of upstream data.
+     * @return Stream containing repeating upstream entry.
+     */
+    private static <K, V> Stream<UpstreamEntry<K, V>> repeatEntry(
+        UpstreamEntry<K, V> en,
+        PoissonDistribution poissonDistribution) {
+        int count = poissonDistribution.sample();
+
+        return IntStream.range(0, count).mapToObj(i -> en);
+    }
+
+    public static IgniteFunction<Vector, Vector> getProjector(Map<Integer, Integer> mapping) {
+        return v -> {
+            Vector res = VectorUtils.zeroes(mapping.size());
+            mapping.keySet().stream().forEach(locId -> res.set(locId, v.get(mapping.get(locId))));
+
+            return res;
+        };
+    }
+
+    private static <K, V> IgniteBiFunction<K, V, Vector> wrapExtractor(IgniteBiFunction<K, V, Vector> featureExtractor, Map<Integer, Integer> featureMapping) {
+        return featureExtractor.andThen((IgniteFunction<Vector, Vector>) featureValues -> {
+            double[] newFeaturesValues = new double[featureMapping.size()];
+            featureMapping.forEach((localId, featureValueId) -> newFeaturesValues[localId] = featureValues.get(featureValueId));
```

Comment:
I would suggest to make it a method of `Vector` to be able to implement it efficiently for sparse vectors.